src/CoreLib/distributedfunction.fs

```diff
@@ -343,6 +343,115 @@ type internal DistributedFunctionHolder(name:string, capacity:int, executor: (Gu
+                            chainedCTS.Dispose()
+                            semaphore.Release() |> ignore
+                        with 
+                        | ex -> 
+                            Logger.LogF( LogLevel.Warning, fun _ -> sprintf "ExecuteWithTimebudget, Functioon %s, exception in main loop of %A" name ex )
+                            observer.OnError( ex )
+                            chainedCTS.Dispose()
+                            semaphore.Release() |> ignore
+                    else
+                        try 
+                            if useTimeBudget <= elapse then 
+                                let ex = OperationCanceledException(sprintf "ExecuteWithTimebudget: Function %s, semaphore task is entered as cancelled" name)
+                                observer.OnError(ex)
+                            else
+                                let ex = TimeoutException( sprintf "ExecuteWithTimebudget: Function %s, semaphore task is entered in %d ms(budget %d ms)" name elapse timeBudget )
+                                observer.OnError(ex)
+                            chainedCTS.Dispose()
+                            semaphore.Release() |> ignore 
+                        with
+                        | ex -> 
+                            Logger.LogF( LogLevel.Warning, fun _ -> sprintf "ExecuteWithTimebudget, Functioon %s, exception while not enter, of %A" name ex )
+                            observer.OnError( ex )
+                            chainedCTS.Dispose()
+                let cancelledExecute (taskEnter:Task<bool>) = 
+                    try 
+                        chainedCTS.Dispose()
+                        semaphore.Release() |> ignore 
+                    with
+                    | ex -> 
+                        Logger.LogF( LogLevel.Warning, fun _ -> sprintf "ExecuteWithTimebudget, Functioon %s, exception when dispose of %A" name ex )
+                semaTask.ContinueWith( wrapperExecute, TaskContinuationOptions.None ) |> ignore 
+                semaTask.ContinueWith( cancelledExecute, TaskContinuationOptions.OnlyOnCanceled ) |> ignore
+            else
+                try 
+                    executor( jobID, useTimeBudget, input, chainedCTS.Token, observer )
+                    chainedCTS.Dispose()
+                with 
+                | ex -> 
+                    chainedCTS.Dispose()
+                    observer.OnError( ex )
+    override x.ToString() = sprintf "Distributed function %s:%s/%d" name (if Utils.IsNull semaphore then "Unknown" else (capacity-semaphore.CurrentCount).ToString() ) (capacity)
+    /// Cancel all jobs related to this distributed function. 
+    member x.Cancel() = 
+        // This process is responsible for the disposing routine
+        cts.Cancel()
+    /// Can we dispose this job holder?
+    member x.CanDispose() = 
+        // Wait for all job in semaphore to exit
+        if capacityControl then 
+            semaphore.CurrentCount = capacity 
```

Comment:
I see. What's the mechanism to wait for these in-going work to exit? So we can safely call dispose?